server/src/main/java/io/druid/server/initialization/AuthenticatorMapperModule.java

```diff
@@ -125,7 +125,43 @@ public AuthenticatorMapper get()
         authenticatorMap.put(authenticatorName, authenticator);
       }
 
-      return new AuthenticatorMapper(authenticatorMap, authConfig.getEscalatedAuthenticator());
+      return new AuthenticatorMapper(authenticatorMap);
+    }
+  }
+
+  private static void validateProperties(Properties properties, List<String> authenticators)
+  {
+    String escalatorType = properties.getProperty("druid.escalator.type");
+    if (escalatorType == null) {
+      escalatorType = "allowAll";
+    }
+
+    if (authenticators == null) {
+      if (escalatorType.equals("allowAll")) {
+        return;
+      } else {
+        throw new ISE("Using default unsecured configuration with invalid druid.escalator.type [%s]", escalatorType);
+      }
+    }
+
+    if (authenticators.isEmpty()) {
+      throw new IAE("Must have at least one Authenticator configured.");
+    }
+
+    Set<String> authenticatorSet = new HashSet<>();
+    for (String authenticator : authenticators) {
+      if (authenticatorSet.contains(authenticator)) {
+        throw new ISE("Cannot have multiple authenticators with the same name: [%s]", authenticator);
+      }
+      authenticatorSet.add(authenticator);
+    }
+
+    for (String authenticator : authenticators) {
+      String typeProperty = StringUtils.format("druid.auth.authenticator.%s.type", authenticator);
+      if (escalatorType.equals(properties.getProperty(typeProperty))) {
```

Comment:
why is it required that escalatorType must match one of the Authenticator types ?  Also, does this mean Escalator impls and Authenticator impls much have same json type names ?

Reply:
Actually, I think this is a good question -- do we really need to require that escalators match up with an authenticator?

I think there's no reason they should need to match exactly. It's plausible that an escalator doesn't line up totally with an authenticator. Imagine that the authenticator is a standard RBAC authenticator, and the escalator is some custom extension that pulls the internal user password from somewhere more secure than a config file.

Reply:
Whatever authentication scheme an Escalator uses, it has to be supported by some Authenticator that's configured, so I impose that type matching restriction to conceptually link the implementations.

It could function without that restriction, but I don't see much reason to have them differ.

This does mean they must have the same JSON type names, I have a section in the doc that calls out this requirement:

```
The Escalator chosen for this property must have the same type as an Authenticator in `druid.auth.authenticationChain. Authenticator extension implementors must also provide a corresponding Escalator implementation with the same type name if they intend to use a particular authentication scheme for internal Druid communications.
```

Reply:
I see, so technically its not necessary but you're imposing the restriction so that user thinks about Escalator and Authenticator impl together.
I think it would help extension writers if we put some comment on both Authenticator and Escalator interfaces that implementers must ensure that both Escalator and Authenticator implementations must have exactly same json type name. Its easy to miss from that line in the doc alone. thanks.

Reply:
@gianm how did your comment to the time travel ?

Reply:
:magic:

Reply:
@himanshug @gianm I removed the type matching enforcement